internal/reconcilers/cluster.go

```diff
@@ -0,0 +1,154 @@
+// object.
+type hardwareIPGetter interface {
+	NextAvailableHardwareID(ctx context.Context) (string, error)
+	GetHardwareIP(ctx context.Context, uuid string) (string, error)
+}
+
+// clusterReconcileContext implements ReconcileContext by reconciling TinkerbellCluster object.
+type clusterReconcileContext struct {
+	ctx               context.Context
+	tinkerbellCluster *infrastructurev1alpha3.TinkerbellCluster
+	patchHelper       *patch.Helper
+	hardwareIPGetter  hardwareIPGetter
+	cluster           *clusterv1alpha3.Cluster
+	log               logr.Logger
+	client            client.Client
+}
+
+// New builds a context for cluster reconciliation process, collecting all required
+// information.
+//
+// If unexpected case occurs, error is returned.
+//
+// If some data is not yet available, nil is returned.
+func (crcc *ClusterReconcileContextConfig) New(namespacedName types.NamespacedName) (ReconcileContext, error) {
+	crc := &clusterReconcileContext{
+		log:               crcc.Log.WithValues("tinkerbellcluster", namespacedName),
+		ctx:               context.Background(),
+		tinkerbellCluster: &infrastructurev1alpha3.TinkerbellCluster{},
+		hardwareIPGetter:  crcc.HardwareIPGetter,
+		client:            crcc.Client,
+	}
+
+	if err := crc.client.Get(crc.ctx, namespacedName, crc.tinkerbellCluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, nil
+		}
+
+		return nil, fmt.Errorf("getting TinkerbellCluster: %w", err)
+	}
+
+	crc.log = crc.log.WithName(crc.tinkerbellCluster.APIVersion)
+
+	patchHelper, err := patch.NewHelper(crc.tinkerbellCluster, crc.client)
+	if err != nil {
+		return nil, fmt.Errorf("initializing patch helper: %w", err)
+	}
+
+	crc.patchHelper = patchHelper
+
+	cluster, err := util.GetOwnerCluster(crc.ctx, crc.client, crc.tinkerbellCluster.ObjectMeta)
+	if err != nil {
+		return nil, fmt.Errorf("getting owner cluster: %w", err)
+	}
+
+	if cluster == nil {
+		crc.log.Info("OwnerCluster is not set yet. Requeuing...")
+
+		return nil, nil
+	}
+
+	crc.cluster = cluster
+
+	if util.IsPaused(crc.cluster, crc.tinkerbellCluster) {
+		crc.log.Info("TinkerbellCluster or linked Cluster is marked as paused. Won't reconcile")
+
+		return nil, nil
+	}
+
+	return crc, nil
+}
+
+// Reconcile implements ReconcileContext interface by ensuring that all TinkerbellCluster object
+// fields are properly populated.
+func (crc *clusterReconcileContext) Reconcile() error {
+	if crc.tinkerbellCluster.Spec.ControlPlaneEndpoint.Host == "" {
+		id, err := crc.hardwareIPGetter.NextAvailableHardwareID(crc.ctx)
+		if err != nil {
+			return fmt.Errorf("getting next available hardware: %w", err)
+		}
+
+		ip, err := crc.hardwareIPGetter.GetHardwareIP(crc.ctx, id)
+		if err != nil {
+			return fmt.Errorf("getting hardware IP: %w", err)
+		}
+
+		crc.log.Info("Assigning IP to cluster", "ip", ip, "clusterName", crc.tinkerbellCluster.Name)
+
+		crc.tinkerbellCluster.Spec.ControlPlaneEndpoint.Host = ip
```

Comment:
This feels a bit racy if we are creating multiple Clusters at the same time, how do we ensure that the id/ip we find here is the one that is used for the initial Machine instance that is created?

Reply:
Yeah, I'm aware of that. We could do some sort of reservation here via cluster objects annotation or to create some kind of stub workflow here to mark hardware as reserved. This might have side-effects though.

>how do we ensure that the id/ip we find here is the one that is used for the initial Machine instance that is created?

It's not completely random, though definitely could be improved. When we deploy controlplane node, we select hardware using this IP address, to make sure the right hardware is picked. So in case new hardware shows up in the meanwhile, we will still select the right one. But there is no protection from other cluster stealing the hardware with this IP...

Having a list of hardware UUIDs for controlplane should improve that.